capstone/capapi/views.py

```diff
@@ -56,19 +57,111 @@ class CitationViewSet(viewsets.GenericViewSet, mixins.RetrieveModelMixin, mixins
 
 
 class CaseViewSet(viewsets.GenericViewSet, mixins.RetrieveModelMixin, mixins.ListModelMixin,):
-    """
-    Browse all cases
-    """
     serializer_class = serializers.CaseSerializer
     http_method_names = ['get']
+    lookup_field = 'slug'
     queryset = models.CaseMetadata.objects.all()
     renderer_classes = (renderers.BrowsableAPIRenderer, renderers.JSONRenderer)
-    lookup_field = 'slug'
+    permission_classes = (permissions.IsAPIUser,)
     filter_class = filters.CaseFilter
 
+    def download_many(self):
+        cases = self.queryset.all().order_by('decision_date')
+
+        for backend in list(self.filter_backends):
+            cases = backend().filter_queryset(self.request, cases, self)
+
+        # get max number of cases to download from query
+        # set max number to daily max if unspecified
+        cases = cases[:int(self.request.query_params.get('max', settings.API_CASE_DAILY_ALLOWANCE))]
+        cases.select_related('jurisdiction')
+        # user is requesting a zip but there is nothing to zip, so 404 is the right response.
+        # See https://stackoverflow.com/a/11760249/307769
+        if cases.count() == 0:
+            return JsonResponse({
+                'message': 'Request did not return any results.'
+            }, status=404, )
+
+        cases = self.paginate_queryset(cases)
+
+        blacklisted_case_count = len(list((case for case in cases if not case.jurisdiction.whitelisted)))
+        case_allowance_sufficient = self.check_case_allowance(blacklisted_case_count)
+
+        response = self.create_download_response(list(cases), blacklisted_case_count, permitted=case_allowance_sufficient)
+
+        return response
+
+    def download_one(self, **kwargs):
+        """
+        If downloading a single case (using its lookup_field) explicitly
+        """
+        try:
+            case = models.CaseMetadata.objects.get(**kwargs)
+        except ObjectDoesNotExist as e:
+            return JsonResponse({'message': 'Unable to find case with matching slug: %s' % e}, status=404, )
+
+        # if whitelisted, count 0, else count 1
+        blacklisted_case_count = int(not case.jurisdiction.whitelisted)
+        case_allowance_sufficient = self.check_case_allowance(blacklisted_case_count)
 
-### User specific views ###
+        response = self.create_download_response([case], blacklisted_case_count, permitted=case_allowance_sufficient)
+
+        return response
+
+    def create_download_response(self, case_list, blacklisted_case_count, permitted=False):
+        if permitted:
+            try:
```

Comment:
We don't want a try-except here, do we? If there's something specific that throws errors we should handle it, otherwise let it throw a 500 so we find out.